transport: add peer information to http2Server and http2Client context

feihu-stripe opened the pull request:
Motivation: https://github.com/grpc/grpc-go/issues/5586

Added peer to the http2server context to allow application to access this information when a connection is just established. This frees application from tracking this information on each and every rpc request.

RELEASE NOTES:
* stats: provide peer information in `HandleConn` context

feihu-stripe commented:
Looks like the failed test case is a known flake test: Test/DetailedGoawayErrorOnGracefulClosePropagatesToRPCError

How can I rerun the build?

feihu-stripe commented:
@easwars 

All tests have passed, but LABEL and MILESTONE I don't have permission to add. Thanks

dfawley commented on internal/transport/http2_server.go:
```diff
@@ -485,14 +488,10 @@ func (t *http2Server) operateHeaders(frame *http2.MetaHeadersFrame, handle func(
-	s.ctx = peer.NewContext(s.ctx, pr)
+
+	// Add peer information to the stream context.
+	s.ctx = peer.NewContext(s.ctx, t.getPeer())
```
Since this is derived from the transport's context (immediately above), we shouldn't need to re-attach the peer here.

feihu-stripe commented:
> Thanks for the PR.
> 
> Did you intend to do this only for the server, and not for the client? If so, why?
> 
> And could you please add tests for this.

Will update PR accordingly

feihu-stripe commented:
Updated:
1. Stopped attaching Peer to stream context. @dfawley 
2. Added a test to validate Peer is set in server context @easwars 

Not feeling confident to make the same change for client due to my limited understand of client side code, also unlike server context which is initiated from [scratch](https://github.com/feihu-stripe/grpc-go/blob/a2e0810efdec0b08a3021281ba838d604dacf1b3/internal/transport/http2_server.go#L246), an [external context](https://github.com/feihu-stripe/grpc-go/blob/a2e0810efdec0b08a3021281ba838d604dacf1b3/internal/transport/http2_client.go#L197) is passed into client constructor, so I am concerned about isolation of that chagne.

easwars commented on internal/transport/transport_test.go:
```diff
@@ -2450,3 +2452,20 @@ func TestConnectionError_Unwrap(t *testing.T) {
 }
+
+// Verify Peer is set in server context.
+func TestPeerSetInServerContext(t *testing.T) {
```
Please specify our test struct as the receiver for this test. This will ensure that our leakcheck runs at the end of the test.

func (s) TestPeerSetInServerContext(t *testing.T) { ... }

easwars commented on internal/transport/transport_test.go:
```diff
@@ -2450,3 +2452,20 @@ func TestConnectionError_Unwrap(t *testing.T) {
+
+// Verify Peer is set in server context.
+func TestPeerSetInServerContext(t *testing.T) {
+	server := setUpServerOnly(t, 0, &ServerConfig{}, suspended)
```
`setUpServerOnly` does not setup any transports. So, `server.conns` in your `for` loop will be `nil`, and therefore this test will not be doing anything.

You either need to use `setUpWithOptions` or you need to explicitly do a `net.Dial()` to create a transport. Please see some other tests in the same file to get a better idea. Thanks.

easwars commented on internal/transport/transport_test.go:
```diff
@@ -2450,3 +2452,20 @@ func TestConnectionError_Unwrap(t *testing.T) {
+	for k := range server.conns {
+		sc, ok := k.(*http2Server)
+		if !ok {
+			t.Fatalf("Failed to convert %v to *http2Server", k)
```
Would be nice to include the actual type as well:

```golang
  t.Fatalf("ServerTransport is of type %T, want %T, k, &http2Server{})
```

easwars commented on internal/transport/transport_test.go:
```diff
@@ -2450,3 +2452,20 @@ func TestConnectionError_Unwrap(t *testing.T) {
+		if !ok {
+			t.Fatalf("Failed to convert %v to *http2Server", k)
+		}
+		_, ok = peer.FromContext(sc.ctx)
+		if !ok {
```
These two lines can be combined into one.

feihu-stripe commented:
@easwars really appreciated for your review.

Will update and respond individually.

dfawley commented:
> an external context is passed into client constructor, so I am concerned about isolation of that chagne.

The main difference is: on the server, the stream's context is a child of the transport's context, whereas in the client, the stream's context is what the user provides for that RPC instead.  It should be fine to attach the peer information on the transport's context as well as the stream's context, but note that we would have to do it in both places on the client.

feihu-stripe commented:
1. set Peer in http2client context
2. in unit test, verify Peer is set in server, client and stream context.

feihu-stripe commented:
r? @easwars @dfawley

easwars commented:
> r? @easwars @dfawley

Could you please take care of the failing tests. Specifically, it is complaining about the following:
```
Error: internal/transport/transport_test.go:2488:4: should replace count += 1 with count++
```

easwars commented on internal/transport/transport_test.go:
```diff
@@ -2450,3 +2452,56 @@ func TestConnectionError_Unwrap(t *testing.T) {
+
+	// verify peer is set in server transport context.
+	count := 0
+	for {
```
You could use the `waitWhileTrue` helper method to simplify the waiting logic here. See : https://github.com/grpc/grpc-go/blob/master/internal/transport/transport_test.go#L1616

feihu-stripe commented:
> Contributor

apologize, missed that error.

feihu-stripe commented:
Is there a recommended way to rerun the CI build? I believe these are unrelated errors.

dfawley commented:
Thank you for the PR!
